### PR TITLE
Update unplug_memory function to enable unplug memdev without dimm

### PR DESCRIPTION
**Pull request by huangyum**
ID: 1725515
Signed-off-by: Yumei Huang <yuhuang@redhat.com>

### CODE REVIEW

**Comment by huangyum** [comment]
(1/3) Host_RHEL.m8.u2.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.2.0.x86_64.io-github-autotest-qemu.hotplug_memory.after.guest_reboot.unplug.buildin_memory.backend_file.policy_interleave.two: PASS (146.59 s)
 (2/3) Host_RHEL.m8.u2.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.2.0.x86_64.io-github-autotest-qemu.hotplug_memory.after.guest_reboot.unplug.pluged_memory.backend_file.policy_interleave.two: PASS (148.80 s)
 (3/3) Host_RHEL.m8.u2.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.2.0.x86_64.io-github-autotest-qemu.hotplug_memory.after.guest_reboot.unplug.unused_memory.backend_file.policy_interleave.two: PASS (129.82 s)

**Comment by huangyum** [comment]
@yanan-fu Would you please help review, thanks.

**Comment by yanan-fu** [comment]
The steps in the comment is not match with the actual steps, the update is correct.
However, I think pre test order is correct: `check mem -> check dimm -> unplug dimm -> unplug mem`.
Because, base on `memory_define_by_params`, if no mem, then no dimm.

The issue is about the `try...expect` when unplug dimm it is a warning when no dimm. Move the unplug part into `try` is ok if i understand correctly.

**Comment by huangyum** [comment]
@yanan-fu Yes, you are right. Just updated the commit summary. Thanks.

**Comment by yanan-fu** [comment]
@huangyum 

> @yanan-fu Yes, you are right. Just updated the commit summary. Thanks.

It is not only about the commit summary, this patch change the order to :
`check dimm -> unplug dimm -> check mem -> unplug mem`, while it should be the pre order.

**Comment by huangyum** [comment]
@yanan-fu Could you pls explain more? What do you mean by 'check mem' ? I didn't change anything about mem.

**Comment by huangyum** [comment]
@yanan-fu I see your point. But I think it's more about 'get' dimm and mem, other than 'check'. The logical is like, get dimm -> unlug dimm -> get mem -> unplug mem, which looks reasonable to me.

**Comment by yanan-fu** [comment]
> @yanan-fu I see your point. But I think it's more about 'get' dimm and mem, other than 'check'. The logical is like, get dimm -> unlug dimm -> get mem -> unplug mem, which looks reasonable to me.

Yes, `get` is more reasonable, that is what i mean.
But, I do not think get dimm before get mem is reasonable, because if no mem then no dimm.

**Comment by huangyum** [comment]
> > @yanan-fu I see your point. But I think it's more about 'get' dimm and mem, other than 'check'. The logical is like, get dimm -> unlug dimm -> get mem -> unplug mem, which looks reasonable to me.
> 
> Yes, `get` is more reasonable, to check if device existing, that is what i mean.
I don't think it's to check if it exists, it just raises error if 'get' fails.
> But, I do not think get dimm before get mem is reasonable, because if no mem then no dimm.
We don't get dimm via mem or mem related info. They are two separate stuff in this function. Get and unlug makes a pair. I don't understand why we have to get mem first. Mem is a must for dimm, but it's not necessary here to get a dimm.

**Comment by yanan-fu** [comment]
> > We don't get dimm via mem or mem related info. They are two separate stuff in this function. Get and unlug makes a pair. I don't understand why we have to get mem first. Mem is a must for dimm, but it's not necessary here to get a dimm.

My point is, for one memory device, if there do not have memory object, no need to test again.  No matter there have dimm or not.

**Comment by huangyum** [comment]
> > > We don't get dimm via mem or mem related info. They are two separate stuff in this function. Get and unlug makes a pair. I don't understand why we have to get mem first. Mem is a must for dimm, but it's not necessary here to get a dimm.
> 
> My point is, for one memory device, if there do not have memory object, no need to test again. No matter there have dimm or not.

The issue discussed here is about the order of 'get mem' and 'unplug dimm'. No matter which one comes first, if it fails, the other one would skip. Do you have other justification?

**Comment by yanan-fu** [comment]
For unplug memory device, it unplug both dimm and mem object.
1. if no dimm device, just a warning, since some mem do not have dimm device since it is not used
2. if no mem, test finished as no mem.

The order you use is: get dimm -> unplug dimm if exist, go ahead if not exist -> check mem -> unplug mem if exist, and test error if no mem.
get dimm failed will not fail the test, it will go ahead to check mem.

My point is, we need get mem first, if no mem, no need to get dimm, test finished:
get mem -> get dimm if mem exist, and test error if no mem - > unplug dimm if exist, go ahead if not exist -> unplug mem.

@huangyum  As we discussed, let's call for more idea about this issue. @luckyh could you help review too, thanks!

File: virttest/utils_test/qemu/__init__.py [hunk 1]
```diff
@@ -634,14 +634,27 @@ def hotplug_memory(self, vm, name):
...
+            dimm = vm.devices.get_by_qid(qid_dimm)[0]
+            step = "Unplug pc-dimm '%s'" % qid_dimm
+            error_context.context(step, logging.info)
+            _, ver_out = vm.devices.simple_unplug(dimm, vm.monitor)
+            if ver_out is False:
+                raise exceptions.TestFail("Verify unplug memory failed")
+            devices.append(dimm)
+            self.update_vm_after_unplug(vm, dimm)
+        except IndexError:
+            logging.warn("'%s' is not used by any dimm" % qid_mem)
```
Comment (luckyh) [inline virttest/utils_test/qemu/__init__.py:656]: Suggest the following code structure:
```python
try:
    dimm = vm.devices.get_by_qid(qid_dimm)[0]
except IndexError:
    logging.warn("'%s' is not used by any dimm" % qid_mem)
else:
    step = "Unplug pc-dimm '%s'" % qid_dimm
    ...
```

Comment (huangyum) [reply virttest/utils_test/qemu/__init__.py]: Thanks, updated. Any more comments?

**Comment by huangyum** [comment]
@yanan-fu Any more comments?

**Comment by yanan-fu** [comment]
> @yanan-fu Any more comments?

I will reserve my opinion, to check the backend first. Otherwise, when test fail as no mem:
1. It waste time to get the dimm.
2. Test log is strange, that a warning that '*qid_mem* is not used by any dimm' and then failed as 'Memory object *qid_mem* not exists'.

But, i am ok if you all can accept this solution, thanks!

**Comment by huangyum** [comment]
> > @yanan-fu Any more comments?
> 
> I will reserve my opinion, to check the backend first. Otherwise, when test fail as no mem:
> 
>     1. It waste time to get the dimm.
> 
I'm ok with this one. Same situation may happen in many places.
>     2. Test log is strange, that a warning that '_qid_mem_ is not used by any dimm' and then failed as 'Memory object _qid_mem_ not exists'.
> 
As to the log,  I can update the log to 'no qid_dimm found' if you insist. Thanks.
> 
> But, i am ok if you all can accept this solution, thanks!

**Comment by yanan-fu** [comment]
> > > @yanan-fu Any more comments?
> > 
> > 
> > I will reserve my opinion, to check the backend first. Otherwise, when test fail as no mem:
> > ```
> > 1. It waste time to get the dimm.
> > ```
> 
> I'm ok with this one. Same situation may happen in many places.
> 
> > ```
> > 2. Test log is strange, that a warning that '_qid_mem_ is not used by any dimm' and then failed as 'Memory object _qid_mem_ not exists'.
> > ```
> 
> As to the log, I can update the log to 'no qid_dimm found' if you insist. Thanks.

As i mentioned in the comment above, i did not insist anything,  just reserve my opinion, i am ok if you and xu all can accept that ~

> 
> > But, i am ok if you all can accept this solution, thanks!

**Comment by huangyum** [comment]
@yanan-fu xu has approved the changes, so I guess he accepts. What do you think to proceed with this patch?

**Comment by yanan-fu** [comment]
> @yanan-fu xu has approved the changes, so I guess he accepts. What do you think to proceed with this patch?

Yes, just go ahead.

**Comment by huangyum** [comment]
@yanan-fu Do you think we need another reviewer, or we can merge directly?

**Comment by yanan-fu** [comment]
> @yanan-fu Do you think we need another reviewer, or we can merge directly?

@huangyum Sorry, I do not have permission to merge avocado-vt patch, Just confirm with xuhan@  please, thanks!

**Comment by huangyum** [comment]
@luckyh Could you please help merge if it's ok, thanks.

**Comment by yanan-fu** [comment]
Ack
@luckyh could you help review again, thanks!

**Comment by luckyh** [comment]
Thanks all.

**Pull request merged**